Add CONF_STATE_CLASS to `sensor/__init__.py`

farmio opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add CONF_STATE_CLASS to `homeassistant/components/sensor/__init__.py`
and use it throughout the project

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

farmio committed:
add CONF_STATE_CLASS to const.py

janiversen reviewed (commented):
I like the idea of having CONF_STATE_CLASS in homeassistant/const.py, but I am not so sure it is a good idea to "hide" it in a PR that also changes integrations.

And btw. you have submitted other PRs (at least one for modbus) where you add it locally, would be nice to put those on hold, in order not to make double changes.

janiversen commented on homeassistant/const.py:
```diff
@@ -181,6 +181,7 @@
 CONF_SOURCE: Final = "source"
 CONF_SSL: Final = "ssl"
 CONF_STATE: Final = "state"
+CONF_STATE_CLASS: Final = "state_class"
```
Is state_class for all types of entities ? my impression is that it is only for sensors, and thus should be defined in sensors/__init__.py

farmio replied:
moved it there.

farmio commented:
I replaced existing integration constants here because this was requested before in a comparable PR adding a validator to core. 
https://github.com/home-assistant/core/pull/41640#issuecomment-706701358

I'll set the Modbus PR to draft meanwhile.

farmio committed:
move to `sensor/__init__.py`

janiversen reviewed (approved):
LGTM.

janiversen reviewed (approved):
LGTM. Thanks.

frenck commented on homeassistant/components/sensor/__init__.py:
```diff
@@ -66,6 +66,7 @@
 
 DEVICE_CLASSES_SCHEMA: Final = vol.All(vol.Lower, vol.In(DEVICE_CLASSES))
 
+CONF_STATE_CLASS: Final = "state_class"
```
Maybe introduce a const.py file instead.

farmio replied:
Should the move to const.py then be done in a separate PR?

frenck replied:
Why? It is added here?

farmio replied:
Can't follow. Do you want to have only conf_state_class in const.py or all constants used in climate/__init__.py?

frenck replied:
This PR adds a new constant... for global use... right?
I'm suggesting to introduce a const.py file and put this one into that file instead.

But as said below in the review comment, before considering doing that, I would like a sign-off from @MartinHjelmare

MartinHjelmare replied:
We normally only define constants for config keys in entity integrations when the constants  are used in the integration itself, like in scene and legacy device_tracker. So I'm a bit hesitant to add it here. If we do add it in the integration I think a `const.py` module is a good idea (option 1). The other option is to add it to the the global `const.py` module since it's a config option that is used by a couple of integrations already (option 2).

I don't have a strong opinion either way. I'd go with option 1, right now.

frenck reviewed (changes requested):
Leaving this as a review, as I think we should consider using const.py files instead of putting them into here.

Tagging/asking @MartinHjelmare for a second opinion on that.

janiversen commented:
Most of the information integrations import from the generic entities (sensor etc.) seems to be defined in the __init__.py

Having a const.py is not a bad idea, but lets please lets not ask the integrations to have 2 import statement from the same entity….all “exported” information in one place please.

frenck commented:
> but lets please lets not ask the integrations to have 2 import statement from the same entity….all “exported” information in one place please.

That can be a nice followup PR

farmio commented:
So is there any action required? I think this is still useful even after the latest changes to state_class (total_increasing).

farmio committed:
Merge remote-tracking branch 'upstream/dev' into conf_state_class

farmio committed:
move to sensor/const.py

farmio commented:
So I've added a `sensor/const.py` file and moved CONF_STATE_CLASS there. This means integrations have an additional import now. I think Martins 2nd option (move to global const.py) would be nicer, but your call!

MartinHjelmare commented:
If we want to avoid the extra import line we can import the names into the `__init__.py` module as unused imports. Again, I could see us go with option 2 also. I don't think it really matters much.

I hope this constant won't be used much since its usage means that the user needs to configure entity behavior which is something we're trying to avoid.

farmio commented:
Wouldn't mypy complain about `implicit-reexport` then? `--no-implicit-reexport` is part of `--strict`.

farmio committed:
Revert "move to sensor/const.py"

This reverts commit 604d0d066bfcb93f1a11e3d7732d430ab6de8d59.

farmio committed:
move it to `sensor/const.py` but import it from `sensor/__init__.py`

farmio committed:
Merge remote-tracking branch 'upstream/dev' into conf_state_class

farmio committed:
add Modbus

MartinHjelmare reviewed (approved):
Thanks!
